File: web/api/models/verse.py
from __future__ import annotations
from typing import List, Optional
from pydantic import BaseModel, ConfigDict


class VersePoint(BaseModel):
    verse_id: int
    book_name: str
    chapter_number: int
    verse_number: int
    x: Optional[float] = None
    y: Optional[float] = None
    z: Optional[float] = None

    model_config = ConfigDict(arbitrary_types_allowed=True)


class WordAlignment(BaseModel):
    word_position: int
    original_word: str
    transliteration: Optional[str] = None
    english_gloss: str
    strongs_number: Optional[str] = None
    morphology_code: Optional[str] = None
    root_definition: Optional[str] = None

    model_config = ConfigDict(arbitrary_types_allowed=True)


class CrossRefBrief(BaseModel):
    target_verse_id: int
    target_ref: str
    relevance_score: float
    text_preview: Optional[str] = None
    x: Optional[float] = None
    y: Optional[float] = None
    z: Optional[float] = None

    model_config = ConfigDict(arbitrary_types_allowed=True)


class TopicRef(BaseModel):
    topic: str

    model_config = ConfigDict(arbitrary_types_allowed=True)


class VerseDetail(BaseModel):
    verse_id: int
    book_name: str
    book_abbreviation: str
    chapter_number: int
    verse_number: int
    text: str
    testament: str
    word_alignments: List[WordAlignment]
    cross_references: List[CrossRefBrief]
    nave_topics: List[TopicRef]

    model_config = ConfigDict(arbitrary_types_allowed=True)

File: web/src/types/verse.ts
export interface VersePoint {
  verseId: number;
  x: number;
  y: number;
  z: number;
  bookId: number;
  chapterNumber: number;
  verseNumber: number;
  crossRefCount: number;
  testament: 'OT' | 'NT';
  genre: string;
  ethicsMaxScore: number;
}

export interface VerseDetail {
  verseId: number;
  bookName: string;
  bookAbbreviation: string;
  chapterNumber: number;
  verseNumber: number;
  testament: string;
  text: string;
  crossRefCount: number;
  topics: string[];
  wordAlignments: WordAlignment[];
  crossReferences: CrossRefBrief[];
}

export interface WordAlignment {
  wordPosition: number;
  originalWord: string;
  transliteration: string;
  englishGloss: string;
  strongsNumber: string;
  morphologyCode: string | null;
  rootDefinition: string;
}

export interface CrossRefBrief {
  verseId: number;
  reference: string;
  relevanceScore: number;
  textPreview: string;
}
